runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -4308,8 +4308,10 @@ class INTERPRETER_CLASS
 			J9TranslationBufferSet *translationData = _vm->dynamicLoadBuffers;
 			if (NULL != translationData) {
 				/* Initialize the class path entry */
+				omrthread_monitor_enter(classLoader->cpEntriesMutex);
 				J9ClassPathEntry *cpEntry = classLoader->classPathEntries + cpIndex;
 				type = (I_32)initializeClassPathEntry(_vm, cpEntry);
+				omrthread_monitor_exit(classLoader->cpEntriesMutex);
```

Comment:
Seems overkill to protect initializeClassPathEntry(), it was never protected before. This api goes to the file system and may be slow, holding the mutex for too long.

Reply:
I agree wrt the time on the filesystem, but "it wasn't protected before" is precisely why this PR exists, so is not much of an argument. Also, the mutex is in the wrong place here - it really needs to be before the range check on classPathEntryCount if it's done at all.

Reply:
If we want to properly protect the entries, there are a lot more changes to make.

Reply:
I don't like the idea of a partial fix - if we defer the rest until later, a lot of the knowledge gained here will need to be relearned.

Reply:
The entries only get freed when the JVM shuts down, do you know of some reason why the entries need to be protected? It's only the array holding them that is the current problem.

Reply:
Are the entries pointed to by the array, or are they packed as part of the array? If the second, then they need to be protected while in use.

Reply:
Ok I see, you are right, they are part of the array. So more work to do.

Reply:
An alternative is to allocate them separately, which is also more work, but perhaps less risky than trying to protect them.

Reply:
> it really needs to be before the range check on classPathEntryCount if it's done at all

Originally I was acquiring the mutex before the check on classPathEntryCount, then I change it to the current place, because  classLoader->classPathEntryCount can only be increased, never decreased. so the mutex won't affect the result of
` (cpIndex < (I_32)classLoader->classPathEntryCount)`

Reply:
I think that's a bit too subtle to gloss over - it would need a comment, but really, we're talking about adding 2 lines of code into the protection, where the call is already expected to take a relatively large amount of time.

Reply:
Separate allocation would certainly solve some of the issues (e.g. locked time spent accessing the filesystem). The reallocated arrays would be smaller in size, which might result in a bit more malloc fragmentation.

Reply:
All the initial bootstrap entries could be allocated together, with only jars added later being separate. Also the initial array could be allocated to hold extra entries and grown by an increment, resulting in fewer allocations or none.

Reply:
I was thinking something similar - entries / maxSize/ usedSize. Allocating the initial entries together might make freeing them problematic (assuming we go with the pointer array solution to avoid the need for locking around the filesystem access).

Reply:
> All the initial bootstrap entries could be allocated together, with only jars added later being separate. 

Which means for appended entries in separate memory, classPathEntries[i] won't work anymore ?

Reply:
It would still work, the array changes to an array of pointers rather than an array of structs. For Java 8 there could be say 20 entries we know about up front, all these entries could be allocated in one big block and the pointers in the array assigned to each individual entry.

Reply:
OK, then as GAC mentioned, when freeing them, we need to know from which entry, it is a separate piece of memory. Probably compare abs((UDATA)classPathEntries[i] - (UDATA)classPathEntries[i-1]) against MAX_ENTRIES_MEM_SIZE.

Reply:
One approach is to store number of initial entries in the J9JavaVM struct.

Reply:
Or do some tricks like store it in the first element of the array, but store and pass the address of the first entry. The code to free would know to look earlier in the array for the number of initial elements and the pointer to free.

Reply:
There could be more than one additional memory pieces. From the test cases in https://github.com/eclipse-openj9/openj9/issues/13414, they are appending more than 700 entries to bootstrap class path.

Reply:
That's ok. The first chunk is allocated together and then the 700 additions are all separate. The initial array size could be 100, and then grow by 100 (for example) each time it's out of space so it doesn't need to allocate a new array for every new entry.